File: IndicatorsUI/MainUI/angular-app/src/app/shared/classes/trendmarker-label-provider.ts
import { TrendMarker, TrendMarkerLabel } from '../../typings/FT';
import { PolarityIds } from '../constants';

export class TrendMarkerLabelProvider {
    trendMarkerToLabel: Map<TrendMarker, TrendMarkerLabel>[] = [];
    trendMarkerLabels: TrendMarkerLabel[] = [];
    cannotBeCalculated: TrendMarkerLabel = { id: TrendMarker.CannotBeCalculated, text: 'Cannot be calculated' };
    noChange: TrendMarkerLabel = { id: TrendMarker.NoChange, text: 'No significant change' };

    constructor(private polarityId: number) {
        switch (this.polarityId) {
            case PolarityIds.RAGLowIsGood:
                this.trendMarkerLabels.push(this.cannotBeCalculated);
                this.trendMarkerLabels.push({ id: TrendMarker.Increasing, text: 'Increasing and getting worse' });
                this.trendMarkerLabels.push({ id: TrendMarker.Decreasing, text: 'Decreasing and getting better' });
                this.trendMarkerLabels.push(this.noChange);
                break;

            case PolarityIds.RAGHighIsGood:
                this.trendMarkerLabels.push(this.cannotBeCalculated);
                this.trendMarkerLabels.push({ id: TrendMarker.Increasing, text: 'Increasing and getting better' });
                this.trendMarkerLabels.push({ id: TrendMarker.Decreasing, text: 'Decreasing and getting worse' });
                this.trendMarkerLabels.push(this.noChange);
                break;

            default:
                this.trendMarkerLabels.push(this.cannotBeCalculated);
                this.trendMarkerLabels.push({ id: TrendMarker.Increasing, text: 'Increasing' });
                this.trendMarkerLabels.push({ id: TrendMarker.Decreasing, text: 'Decreasing' });
                this.trendMarkerLabels.push(this.noChange);
                break;
        }

        this.trendMarkerLabels.forEach(trendMarkerLabel => {
            this.trendMarkerToLabel[trendMarkerLabel.id.toString()] = trendMarkerLabel;
        });
    }

    getLabel(trendMarker: number): string {
        return this.trendMarkerLabels[trendMarker].text;
    }
}

File: PholioVisualisationWS/Formatting/TrendMarkerLabelProvider.cs
using System.Collections.Generic;
using System.Linq;
using PholioVisualisation.PholioObjects;

namespace PholioVisualisation.Formatting
{
    public class TrendMarkerLabelProvider
    {
        private readonly Dictionary<TrendMarker, TrendMarkerLabel> _trendMarkerToLabel;
        public readonly IList<TrendMarkerLabel> Labels;

        public TrendMarkerLabelProvider(int polarityId)
        {
            switch (polarityId)
            {
                case PolarityIds.RagLowIsGood:
                    Labels = new List<TrendMarkerLabel> {
                        CannotBeCalculated,
                        new TrendMarkerLabel{Id = TrendMarker.Increasing, Text = "Increasing and getting worse"},
                        new TrendMarkerLabel{Id = TrendMarker.Decreasing, Text = "Decreasing and getting better"},
                        NoChange
                    };
                    break;

                case PolarityIds.RagHighIsGood:
                    Labels = new List<TrendMarkerLabel> {
                        CannotBeCalculated,
                        new TrendMarkerLabel{Id = TrendMarker.Increasing, Text = "Increasing and getting better"},
                        new TrendMarkerLabel{Id = TrendMarker.Decreasing, Text = "Decreasing and getting worse"},
                        NoChange
                    };
                    break;

                default:
                    Labels = new List<TrendMarkerLabel> {
                        CannotBeCalculated,
                        new TrendMarkerLabel{Id = TrendMarker.Increasing, Text = "Increasing"},
                        new TrendMarkerLabel{Id = TrendMarker.Decreasing, Text = "Decreasing"},
                        NoChange
                    };
                    break;
            }

            _trendMarkerToLabel = Labels.ToDictionary(x => x.Id, x => x);
        }

        public TrendMarkerLabel GetLabel(TrendMarker trendMarker)
        {
            return _trendMarkerToLabel[trendMarker];
        }

        private TrendMarkerLabel CannotBeCalculated
        {
            get
            {
                return new TrendMarkerLabel
                {
                    Id = TrendMarker.CannotBeCalculated, Text = "Cannot be calculated"
                };
            }
        }

        private TrendMarkerLabel NoChange
        {
            get
            {
                return new TrendMarkerLabel
                {
                    Id = TrendMarker.NoChange, Text = "No significant change"
                };
            }
        }
    }
}